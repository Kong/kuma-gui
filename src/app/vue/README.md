---
section: Packages
---
# vue

Automatic `vue` initialization using our projects dependency injection layer.

