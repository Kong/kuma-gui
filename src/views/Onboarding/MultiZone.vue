<template>
  <OnboardingPage>
    <template #header>
      <OnboardingHeading title="Adding a new zone" />
    </template>
    <template #content>
      <p class="text-center mb-4">
        In a multi-zone deployment, in order to start our service mesh we must first add a new zone to our deployment by
        starting a Zone CP and its Ingress/Egress.
      </p>

      <p class="text-center my-8">
        A Zone CP will ultimately receive connections from the data plane proxies in the same zone, therefore it needs
        to be configured in such a way that this connectivity is possible. There is no strict rule to how big or small a
        zone can be as long as this requriement is satisifed. Therefore, a Zone CP can be a Kubernetes cluster, a VPC, a
        cloud or region.
      </p>

      <CodeView
        title="To get started adding a new Zone, please follow the steps from link below:"
        copy-button-text="Copy link"
        lang="bash"
        :content="documentationLink"
      />
    </template>

    <template #navigation>
      <OnboardingNavigation next-step="onboarding-multi-zone-status" previous-step="onboarding-backend-types" />
    </template>
  </OnboardingPage>
</template>

<script>
import OnboardingNavigation from '@/views/Onboarding/components/OnboardingNavigation'
import OnboardingHeading from '@/views/Onboarding/components/OnboardingHeading'
import OnboardingPage from '@/views/Onboarding/components/OnboardingPage'
import CodeView from '@/components/Skeletons/CodeView'

export default {
  name: 'Multizone',
  components: {
    OnboardingNavigation,
    OnboardingHeading,
    OnboardingPage,
    CodeView,
  },
  metaInfo() {
    return {
      title: 'Multizone',
    }
  },
  data() {
    return {
      documentationLink: 'https://kuma.io/docs/latest/deployments/multi-zone/#zone-control-plane',
    }
  },
}
</script>
