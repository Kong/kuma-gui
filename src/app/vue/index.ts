import { createApp } from 'vue'
import {
  createRouter,
  createWebHistory,
} from 'vue-router'

import type { ServiceDefinition } from '@/services/utils'
import { token, createInjections } from '@/services/utils'
import type { Component } from 'vue'
import type { Router, RouteRecordRaw, NavigationGuard } from 'vue-router'

type Token = ReturnType<typeof token>

type VueApp = ReturnType<typeof createApp>
export type PluginDefinition = Parameters<VueApp['use']>
export type ComponentDefinition = [string, Component]

const $ = {
  app: token<(App: Component) => Promise<VueApp>>('vue.app'),
  router: token<Router>('vue.router'),

  components: token('vue.components'),
  plugins: token<PluginDefinition[]>('vue.plugins'),
  routes: token<RouteRecordRaw[]>('vue.routes'),
  routesLabel: token<RouteRecordRaw[]>('vue.routes.label'),
  navigationGuards: token<NavigationGuard[]>('vue.routes.navigation.guards'),
}
export const services = (app: Record<string, Token>): ServiceDefinition[] => {
  return [
    [$.app, {
      service: (
        components: ComponentDefinition[],
        plugins: PluginDefinition[],
      ) => {
        return async (App: Component) => {
          const app = createApp(App)

          plugins.forEach(([...args]) => {
            app.use(...args)
          })

          components.forEach(([name, component]: [string, Component]) => {
            app.component(name, component)
          })

          return app
        }
      },
      arguments: [
        $.components,
        $.plugins,
      ],
    }],
    [$.router, {
      service: (
        env: (str: 'KUMA_BASE_PATH') => string,
        routes: RouteRecordRaw[],
        guards: NavigationGuard[],
      ) => {
        const router = createRouter({
          history: createWebHistory(env('KUMA_BASE_PATH')),
          routes,
        })

        guards.forEach((item) => {
          if (typeof item === 'function') {
            router.beforeEach(item)
          }
        })

        return router
      },
      arguments: [
        app.env,
        $.routes,
        $.navigationGuards,
      ],
    }],

    [token('vue.plugins'), {
      service: (router) => {
        return [
          [router],
        ]
      },
      arguments: [
        $.router,
      ],
      labels: [
        $.plugins,
      ],
    }],

    // TODO(jc): Label decoration. Make it so we can decorate labels
    // Temporarily turn the routesLabel label into routes for decoration
    // purposes
    [$.routes, {
      service: (routes: RouteRecordRaw[]) => routes,
      arguments: [
        $.routesLabel,
      ],
    }],
    //

    // TODO(jc): Make it so we don't need to provide a [] empty default for
    // labels
    [token('application.routes.navigation.guards'), {
      service: () => {
        return []
      },
      labels: [
        $.navigationGuards,
      ],
    }],
  ]
}

export const TOKENS = $
export const [
  useRouter,
] = createInjections(
  TOKENS.router,
)
