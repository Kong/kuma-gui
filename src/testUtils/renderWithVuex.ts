import { render, RenderOptions, RenderResult } from '@testing-library/vue'
import merge from 'lodash/merge'
import { StoreOptions } from 'vuex'
import Vue, { ComponentOptions } from 'vue'
import Store from '@/store'

function renderWithVuex(
  Component: typeof Vue | ComponentOptions<Vue>,
  options: RenderOptions<Vue> = { store: {}, routes: [] },
): RenderResult {
  const store: StoreOptions<any> = Store()

  // Render the component and merge the original store and the custom one
  // provided as a parameter. This way, we can alter some behaviors of the
  // initial implementation.

  return render(Component, { ...options, store: merge(store, options.store), routes: options.routes })
}

export default renderWithVuex
