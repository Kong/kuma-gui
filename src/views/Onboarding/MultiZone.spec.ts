import renderWithVuex from '@/testUtils/renderWithVuex'
import { screen } from '@testing-library/vue'

import Kuma from '@/services/kuma'
import MultiZone from './MultiZone.vue'

describe('MultiZone.vue', () => {
  it('renders snapshot', () => {
    const { container } = renderWithVuex(MultiZone)

    expect(container).toMatchSnapshot()
  })
  // TODO before merge remove skip
  xit('detects resources on call and allow to proceed', async () => {
    renderWithVuex(MultiZone, {
      routes: [],
    })

    expect(await screen.findByTestId('zone-online')).toBeInTheDocument()
    expect(await screen.findByTestId('zone-ingress-online')).toBeInTheDocument()
    expect(screen.queryByText(/Next/)).toBeInTheDocument()
    expect(screen.queryByTestId('loading')).not.toBeInTheDocument()
  })

  it('refetch resources if any not available', async () => {
    jest.useFakeTimers()
    jest
      .spyOn(Kuma, 'getAllZoneIngressOverviews')
      .mockReturnValueOnce({
        // @ts-ignore
        total: 0,
      })
      .mockReturnValueOnce({
        // @ts-ignore
        total: 1,
      })

    renderWithVuex(MultiZone, {
      routes: [],
    })

    expect(await screen.findByTestId('zone-online')).toBeInTheDocument()
    expect(screen.queryByTestId('loading')).toBeInTheDocument()
    expect(screen.queryByTestId('zone-ingress-offline')).toBeInTheDocument()

    jest.runAllTimers()

    expect(Kuma.getAllZoneIngressOverviews).toHaveBeenCalledTimes(2)

    expect(await screen.findByTestId('zone-ingress-online')).toBeInTheDocument()
  })
})
