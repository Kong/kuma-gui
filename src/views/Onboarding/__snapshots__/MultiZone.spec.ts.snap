// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MultiZone.vue renders snapshot 1`] = `
<div>
  <div>
    <div
      class="onboarding-container"
    >
      <div
        class="onboarding-container__header"
      >
        <div
          class="relative"
        >
          <h1
            class="onboarding-title"
          >
            
    Add zones
  
          </h1>
           
          <!---->
        </div>
      </div>
       
      <div
        class="onboarding-container__content"
      >
        <div
          class="w-full"
        >
          <p
            class="text-center mb-4"
          >
            
      A zone requires both the zone control plane and zone ingress. On Kubernetes, you run a single command to create both resources. On Universal, you must create them separately.
    
          </p>
           
          <div
            class="code-view"
          >
            <div
              class="code-view-content"
            >
              <div
                class="kong-card noBorder"
                data-v-0029352b=""
              >
                <div
                  class="k-card-header mb-4"
                  data-v-0029352b=""
                >
                  <div
                    class="k-card-title"
                    data-v-0029352b=""
                  >
                    <h4
                      data-v-0029352b=""
                    >
                      See the documentation for options to install:
                    </h4>
                  </div>
                  <div
                    class="k-card-actions"
                    data-v-0029352b=""
                  >
                    <div
                      data-v-0029352b=""
                      data-v-3cb990b6=""
                    >
                      <button
                        class="k-button primary"
                        data-v-38fa8ecf=""
                        data-v-3cb990b6=""
                        type="button"
                      >
                        
              Copy link
            
                      </button>
                      <transition-stub
                        data-v-3cb990b6=""
                        name="fade"
                      >
                        <div
                          class="k-popover"
                          data-v-3cb990b6=""
                          style="width: 200px; display: none;"
                        >
                          <!---->
                          <div
                            class="k-popover-content"
                            data-v-3cb990b6=""
                          >
                            <div>
                              <p>
                                Entity copied to clipboard!
                              </p>
                            </div>
                          </div>
                        </div>
                      </transition-stub>
                    </div>
                  </div>
                </div>
                <!---->
                <div
                  class="k-card-body"
                  data-v-0029352b=""
                >
                  <pre
                    class="code-block"
                    data-v-0029352b=""
                  >
                    <code
                      class="language-bash"
                    />
                  </pre>
                </div>
              </div>
            </div>
             
            <div>
              <!---->
               
              <!---->
               
              <!---->
            </div>
          </div>
           
          <div>
            <p
              class="text-center my-4"
            >
              
        Zone status:
        
              <span
                class="text-red-500"
                data-testid="zone-disconnected"
              >
                Disconnected
              </span>
            </p>
             
            <p
              class="text-center mt-4 mb-10"
            >
              
        Zone ingress status:
        
              <span
                class="text-red-500"
                data-testid="zone-ingress-disconnected"
              >
                Disconnected
              </span>
            </p>
             
            <div
              class="flex justify-center"
            >
              <div
                class="loading"
                data-testid="loading"
              >
                <div />
                 
                <div />
                 
                <div />
                 
                <div />
                 
                <div />
              </div>
            </div>
          </div>
        </div>
      </div>
       
      <div
        class="mt-4 flex items-center flex-col sm:flex-row justify-between"
      >
        <a
          aria-current="page"
          class="k-button navigation-button navigation-button--back router-link-exact-active router-link-active primary"
          data-v-38fa8ecf=""
          href="#/"
          type="button"
        >
          
    Back
  
        </a>
         
        <!---->
      </div>
    </div>
     
    <div
      class="background-image"
    />
  </div>
</div>
`;
